rfcs/0024-Non-consensus-node-protocol.md

```diff
@@ -0,0 +1,224 @@
+
+2. Randomsub: For every message to be sent, select upto a certain number of peers randomly. The message propagation would be non-deterministic here but that's ok?
+
+3. Meshsub: Instead of selecting peers randomly for each message to be sent, the node would maintain a mesh (whenever a peer A is in the mesh of peer B, then peer B is also in the mesh of peer A) of randomly selected peers and broadcast all the messages to peers in that set making it somewhat more deterministic than Randomsub. When a new node joins it selects another random set of peers and adds it to the mesh
+
+A note from the go pubsup library (https://github.com/libp2p/go-libp2p-pubsub) : Subscription is not an instanteneous operation. It may take some time before the subscription is processed by the pubsub main loop and propagated to our peers.
+
+### Staged-ledger-validity
+
+When a node verifies a blockchain snark, it also verifies the ledger state upto a certain sequence of transactions (all the transactions for which there is a ledger proof). The transactions that haven't been completely snarked yet, full nodes apply them to the ledger in staged ledger and verify the staged-ledger-hash that is part of the new block is equal to the one they generated. This however is not possible for non-consensus nodes because they don't have any state apart from the two different block headers. We can make this slightly better for the non-consensus nodes by having the block porducers commit to the transactions being included and having an in-snark check for preventing any possible transaction reversal
+
+This can be done by chaining the ledger hash (of the ledger in the staged ledger) like we currently do for coinbases in the pending coinbase data structure.
+
+The stack operations would now be:
+
+    ``` ocaml
+        module Stack : sig
+            val push : t -> Coinbase.t option -> Frozen_ledger_hash.t -> t
+
+            module Checked : sig
+                val push : t -> Coinbase_data.var option -> Frozen_ledger_hash.var -> (t, 'a) Tick.Checked.t
+            end
+        end
+    ```
+The blockchain snark would then prove that the transactions included in a new ledger proof is infact what is expected by checking that the stack corresponding to the new ledger proof is the same as the one updated in the pending coinbase.
+
+### Shared Libraries
+
+Binary for non-consensus nodes can share the following libraries with the one for full-node:
+
+    1. Gossip net
+    2. Verifier (This needs fixing as it currently loads prover keys as well)
+    3. Snarky libraries (update and reuse snarkette?)
+    4. Best tip prover
+    5. Network pool (for transactions)
+    6. `select` function from the consensus library
+    7. Account, Transaction, user command, blockchain state, consensus state and all its contents 
+
+### Incentivizing-full-nodes
+
```

Comment:
Aren't you already incentivized for being a full-node? As a full-node, you can propose and can rewarded for doing snark  work.

Reply:
You'd get that even without serving non-consensus nodes? Non-consensus nodes don't contribute towards any requests that full nodes make but make a lot of requests themselves. For example, merkle paths for every new block. So that's a tiny cost there. But like @bkase said, full-nodes are probably willing to pay it